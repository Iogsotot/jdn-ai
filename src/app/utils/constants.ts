import { BaseUrl } from '../types/mainSlice.types';

export const devUrl = 'http://10.253.219.156:5000';
export const prodUrl = 'http://10.253.219.156:80';

export const URL: { [key: string]: BaseUrl } = {
  devRemote: 'http://10.253.219.156:5000',
  prodRemote: 'http://10.253.219.156:80',
  local: 'http://localhost:5050',
};

export const RemoteUrl: BaseUrl = __DEV_ENVIRONMENT__ ? URL.devRemote : URL.prodRemote;

export enum componentsTexts {
  SeveralTabsWarningMessage = 'You can use the plugin only within 1 tab!',
  SeveralTabsWarningDescription = 'The plugin is open in several tabs now. Please leave 1 plugin tab open and continue working in it.',
  SeveralTabsWarningRetryButton = 'Retry',
  StatusBarVersionJdn = 'JDN v',
  StatusBarVersionBackend = 'Back-end v',
  StatusBarVersionReadme = 'Readme',
  StatusBarServerNoConnection = 'No connection',
  StatusBarLocalServer = 'Local server',
  StatusBarRemoteServer = 'Remote server',
  OnboardingButtonTitle = 'Onboarding tutorial',
}
