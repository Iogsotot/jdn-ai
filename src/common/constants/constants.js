export const readmeLinkAddress = "https://github.com/jdi-testing/jdn-ai#readme";

export const identificationStatus = {
  noStatus: "",
  loading: "Loading...",
  success: "Successful!",
  removed: "Removed",
  error: "An error occurred",
};

export const locatorsGenerationStatus = {
  noStatus: "",
  started: "Locators generation is running in background...",
  complete: "Locators generation is successfully completed",
  failed: "Network error, check network settings or run server locally.",
};

export const locatorProgressStatus = {
  PENDING: "PENDING",
  STARTED: "STARTED",
};

export const locatorTaskStatus = {
  FAILURE: "FAILURE",
  RECEIVED: "RECEIVED",
  REVOKED: "REVOKED",
  RETRY: "RETRY",
  SUCCESS: "SUCCESS",
  ...locatorProgressStatus,
};

export const VALIDATION_ERROR_TYPE = {
  DUPLICATED_NAME: "DUPLICATED_NAME",
  DUPLICATED_LOCATOR: "DUPLICATED_LOCATOR", // warn
  INVALID_NAME: "INVALID_NAME",
  EMPTY_VALUE: "EMPTY_VALUE",
  MULTIPLE_ELEMENTS: "MULTIPLE_ELEMENTS", // warn
  NEW_ELEMENT: "NEW_ELEMENT", // success
  NOT_FOUND: "NOT_FOUND", // warn
};

export const pageType = {
  pageObject: "pageObject",
  locatorsList: "locatorsList",
};

export const copyTitle = {
  Copy: "Copy",
  Copied: "Copied",
};

export const SCRIPT_ERROR = {
  NO_RESPONSE: "The message port closed before a response was received.",
  NO_CONNECTION: "Could not establish connection. Receiving end does not exist.",
};

export const LOCATOR_CALCULATION_PRIORITY = {
  INCREASED: "INCREASED",
  DECREASED: "DECREASED",
};
