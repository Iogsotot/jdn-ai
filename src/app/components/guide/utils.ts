import { ReactElement } from "react";
import { GuideText } from "./text.constants";

export const splitMD = (source: string) => source.match(/^#+ [^#]*/gm);

export const pluginGuide = (splittedMD: Array<string>): string =>
  splittedMD.find((markdown: string) => markdown.includes("### Frontend setup")) || "";
export const serverGuide = (splittedMD: Array<string>) =>
  splittedMD.find((markdown: string) => markdown.includes("### Server setup")) || "";

export const getSteps = (pluginGuideComponent: ReactElement | null, serverGuideComponent: ReactElement | null) => [
  {
    title: "Step 1",
    key: "Step1",
    description: GuideText.FrontendSetup,
    content: pluginGuideComponent,
  },
  {
    title: "Step 2",
    key: "Step2",
    description: GuideText.ServerSetup,
    content: [serverGuideComponent],
  },
  {
    title: "Step 3",
    key: "Step3",
    description: GuideText.Check,
    content: GuideText.CheckDescription,
  },
];

export enum AlertStatus {
  Checking,
  Hide,
  Show,
}
