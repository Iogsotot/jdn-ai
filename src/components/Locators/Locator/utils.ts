import { filter, map } from "lodash";
import { connector } from "../../../services/connector";
import { Locator, LocatorValue } from "../../../store/slices/locatorSlice.types";
import { ElementLibrary } from "../../PageObjects/utils/generationClassesMap";
import { createElementName } from "../../PageObjects/utils/pageObject";

export const getLocator = ({ fullXpath, robulaXpath, customXpath }: LocatorValue) => {
  return customXpath || robulaXpath || fullXpath || "";
};

export const isValidJavaVariable = (value: string) => /^[a-zA-Z_$]([a-zA-Z0-9_])*$/.test(value);

export const evaluateXpath = (xPath: string) => {
  return chrome.storage.sync.set({ xPath }).then(() => {
    return connector.attachContentScript(async () => {
      return await chrome.storage.sync.get(["xPath"]).then(({ xPath }) => {
        try {
          const nodeSnapshot = document.evaluate(xPath, document, null, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null);
          const length = nodeSnapshot.snapshotLength;
          const foundElement = nodeSnapshot.snapshotItem(0) as Element;
          const foundHash = foundElement && foundElement.getAttribute("jdn-hash");
          return JSON.stringify({ length, foundHash });
        } catch (error) {
          return "The locator was not found on the page.";
        }
      });
    });
  });
};

export const equalHashes = (jdnHash: string, locators: Locator[]) => filter(locators, { jdnHash });

export const createNewName = (
    element: Locator,
    newType: string,
    library: ElementLibrary,
    elements: Locator[]
): string => {
  const names = map(elements, "name");
  const newName = createElementName({...element}, library, names, newType);

  return newName;
};
