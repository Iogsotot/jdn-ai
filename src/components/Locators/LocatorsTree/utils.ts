import { cloneDeep, size, map as mapFunction } from "lodash";
import { Locator } from "../../../store/slices/locatorSlice.types";
import { getLocator } from "../Locator/utils";
import { SearchState } from "./LocatorsTree";

export const includesSeacrSubstr = (strings: Array<string>, searchString: string) => {
  const includesSubstring = strings.filter((string) => {
    return string && string.toLowerCase().includes(searchString.toLowerCase());
  });
  return !!size(includesSubstring);
};

export const applySearch = (element: Locator, seacrhString: string): SearchState => {
  const { locator, type, name } = element;
  if (includesSeacrSubstr([getLocator(locator), type as string, name], seacrhString)) {
    return SearchState.None;
  } else return SearchState.Hidden;
};

export interface LocatorTree extends Omit<Locator, "children"> {
  children?: Array<LocatorTree>;
}

export const convertListToTree = (_list: Array<Locator>) => {
  const list: Array<LocatorTree> = mapFunction(cloneDeep(_list), (elem) => ({ ...elem, children: [] }));
  const map: Record<string, number> = {};
  const tree = [];

  for (let i = 0; i < list.length; i++) {
    map[list[i].jdnHash] = i;
  }

  for (let i = 0; i < list.length; i++) {
    const node = list[i];
    if (node.parent_id !== "") {
      const children = list[map[node.parent_id]]?.children;
      children && children.push(node);
    } else {
      tree.push(node);
    }
  }

  return tree;
};
