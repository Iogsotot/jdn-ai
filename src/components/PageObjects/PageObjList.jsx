import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import Icon from "@ant-design/icons";
import { Collapse, Dropdown, Menu, Typography } from "antd";
import { size } from "lodash";

import { PageObjListHeader } from "./PageObjListHeader";

import CaretDownSvg from "../../assets/caret-down.svg";
import PageSvg from "../../assets/page.svg";
import EllipsisSvg from "../../assets/ellipsis.svg";
import TrashBinSvg from "../../assets/trash-bin.svg";
import DownloadSvg from "../../assets/download.svg";
import PencilSvg from "../../assets/pencil.svg";
import { selectConfirmedLocators, selectPageObjects } from "../../store/selectors/pageObjectSelectors";
import { Locator } from "../Locators/Locator";
import { GenerationButtons } from "./GenerationButtons";
import { PageObjectPlaceholder } from "./PageObjectPlaceholder";
import { removePageObject } from "../../store/slices/pageObjectSlice";
import { removeLocators } from "../../store/slices/locatorsSlice";
import { generatePageObject } from "../../services/pageObject";
import { pageType } from "../../utils/constants";
import { changePage } from "../../store/slices/mainSlice";

export const PageObjList = () => {
  const state = useSelector((state) => state);
  const dispatch = useDispatch();
  const currentPageObject = useSelector((state) => state.pageObject.currentPageObject);
  const pageObjects = useSelector(selectPageObjects);
  const [activePanel, setActivePanel] = useState([]);

  useEffect(() => {
    setActivePanel([currentPageObject]);
  }, [currentPageObject]);

  const renderLocators = (elements) => {
    if (size(elements)) {
      return elements.map((element) => <Locator key={element.element_id} {...{ element }} noScrolling={true} />);
    } else {
      return "No locators selected";
    }
  };

  const renderPageObjSettings = (pageObjId, url) => {
    return (
      <div className="jdn__pageObject__settings">
        <div className="jdn__pageObject__settings-url">{url}</div>
        <GenerationButtons pageObj={pageObjId} />
      </div>
    );
  };

  const renderContent = (pageObjId, url, elements) => {
    if (size(elements)) {
      return renderLocators(elements);
    } else {
      return renderPageObjSettings(pageObjId, url);
    }
  };

  const handleRename = (id, name) => {
    chrome.storage.sync.set({ OPEN_EDIT_NAME: { isOpen: true, value: {id, name} } });
  };

  const renderMenu = (id, locatorIds, locatorObjects, name) => {
    const handleRemove = () => {
      dispatch(removePageObject(id));
      dispatch(removeLocators(locatorIds));
    };

    const handleDownload = () => {
      generatePageObject(locatorObjects, name);
    };

    const handleEdit = () => {
      dispatch(changePage({ page: pageType.locatorsList, pageObj: currentPageObject, alreadyGenerated: true }));
    };

    return (
      <Menu>
        <Menu.Item key="3" icon={<PencilSvg />} onClick={() => handleRename(id, name)}>
          Rename
        </Menu.Item>
        {size(locatorIds) ? (
          <React.Fragment>
            <Menu.Item key="4" icon={<PencilSvg />} onClick={handleEdit}>
              Edit list
            </Menu.Item>
            <Menu.Item key="5" icon={<DownloadSvg />} onClick={handleDownload}>
              Download
            </Menu.Item>
          </React.Fragment>
        ) : null}
        <Menu.Item key="6" icon={<TrashBinSvg />} onClick={handleRemove}>
          <Typography.Text type="danger">Delete</Typography.Text>
        </Menu.Item>
      </Menu>
    );
  };

  return (
    <div className="jdn__locatorsList">
      <PageObjListHeader />
      <div className="jdn__locatorsList-content jdn__pageObj-content">
        {size(pageObjects) ? (
          <Collapse
            expandIcon={({ isActive }) => (
              <Icon component={CaretDownSvg} rotate={isActive ? 180 : 270} fill="#808080" />
            )}
            expandIconPosition="left"
            activeKey={activePanel}
            onChange={setActivePanel}
          >
            {pageObjects.map(({ id, name, url, locators }) => {
              const elements = selectConfirmedLocators(state, id);
              return (
                <Collapse.Panel
                  key={id}
                  header={
                    <React.Fragment>
                      <Icon component={PageSvg} className="jdn__locatorsList-status" />
                      {name}
                    </React.Fragment>
                  }
                  extra={
                    <a onClick={(e) => e.stopPropagation()} data-testid="dropdown-button">
                      <Dropdown trigger="click" overlay={renderMenu(id, locators, elements, name)}>
                        <Icon component={EllipsisSvg} />
                      </Dropdown>
                    </a>
                  }
                >
                  {renderContent(id, url, elements)}
                </Collapse.Panel>
              );
            })}
          </Collapse>
        ) : (
          <PageObjectPlaceholder />
        )}
      </div>
    </div>
  );
};
