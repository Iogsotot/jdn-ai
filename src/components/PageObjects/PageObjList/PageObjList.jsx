import React, { useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { Collapse } from "antd";
import { size } from "lodash";
import Icon from "@ant-design/icons";

import { PageObjListHeader } from "./PageObjListHeader";
import { PageObjMenu } from "./PageObjMenu";

import CaretDownSvg from "../../../assets/caret-down.svg";
import PageSvg from "../../../assets/page.svg";
import { selectConfirmedLocators, selectPageObjects } from "../../../store/selectors/pageObjectSelectors";
import { Locator } from "../../Locators/Locator";
import { GenerationButtons } from "./GenerationButtons";
import { PageObjectPlaceholder } from "../PageObjectPlaceholder";
import { PageObjCopyButton } from "./PageObjCopyButton";

export const PageObjList = (props) => {
  const state = useSelector((state) => state);
  const currentPageObject = useSelector((state) => state.pageObject.currentPageObject);
  const pageObjects = useSelector(selectPageObjects);
  const [activePanel, setActivePanel] = useState([currentPageObject]);

  useEffect(() => {
    setActivePanel([currentPageObject]);
  }, [currentPageObject]);

  const renderLocators = (elements) => {
    if (size(elements)) {
      return elements.map((element) => <Locator key={element.element_id} {...{ element }} noScrolling={true} />);
    } else {
      return "No locators selected";
    }
  };

  const renderPageObjSettings = (pageObjId, url) => {
    return (
      <div className="jdn__pageObject__settings">
        <div className="jdn__pageObject__settings-url">{url}</div>
        <GenerationButtons pageObj={pageObjId} />
      </div>
    );
  };

  const renderContent = (pageObjId, url, elements) => {
    if (size(elements)) {
      return renderLocators(elements);
    } else {
      return renderPageObjSettings(pageObjId, url);
    }
  };

  return (
    <div className="jdn__locatorsList">
      <PageObjListHeader {...props} />
      <div className="jdn__locatorsList-content jdn__pageObj-content">
        {size(pageObjects) ? (
          <Collapse
            expandIcon={({ isActive }) => (
              <Icon component={CaretDownSvg} rotate={isActive ? 180 : 270} fill="#808080" />
            )}
            expandIconPosition="start"
            activeKey={activePanel}
            onChange={setActivePanel}
          >
            {pageObjects.map(({ id, name, url, locators, library }) => {
              const elements = selectConfirmedLocators(state, id);
              const isPageObjectNotEmpty = !!size(elements);
              return (
                <Collapse.Panel
                  key={id}
                  header={
                    <React.Fragment>
                      <Icon component={PageSvg} className="jdn__locatorsList-status" />
                      <span className="jdn__pageObject-content-text">{name}</span>
                    </React.Fragment>
                  }
                  extra={
                    <>
                      {isPageObjectNotEmpty && <PageObjCopyButton {...{ elements }} />}
                      <PageObjMenu {...{ id, name, locators, elements, library }} />
                    </>
                  }
                >
                  {renderContent(id, url, elements)}
                </Collapse.Panel>
              );
            })}
          </Collapse>
        ) : (
          <PageObjectPlaceholder />
        )}
      </div>
    </div>
  );
};
