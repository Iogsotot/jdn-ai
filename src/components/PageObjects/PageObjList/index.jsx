import { PageObjList as _PageObjList } from "./PageObjList";

export const PageObjList = _PageObjList;
