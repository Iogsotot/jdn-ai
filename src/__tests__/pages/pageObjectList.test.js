import React from "react";
import * as reactRedux from 'react-redux';
import "@testing-library/jest-dom";
import { jest } from "@jest/globals";
import { render, cleanup } from "@testing-library/react";

import { selectPageObjects } from "../../store/selectors/pageObjectSelectors";
import * as pageObjectSelectors from "../../store/selectors/pageObjectSelectors";
import * as pageObjects from "../__mocks__/pageObjects.mock.json";
import { PageObjList } from "../../components/PageObjects/PageObjList";
import { identificationStatus } from "../../utils/constants";

describe("pageObjectList component", () => {
  const useSelectorMock = jest.spyOn(reactRedux, "useSelector");
  const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");

  const mockSelectors = (selector, store) => {
    if (selector === selectPageObjects) {
      return [pageObjects[0]];
    }
    return selector(store);
  };

  beforeEach(() => {
    useDispatchMock.mockImplementation(() => () => {});
    useSelectorMock.mockImplementation((selector) => mockSelectors(selector, mockStore));

    jest.spyOn(pageObjectSelectors, "selectConfirmedLocators").mockImplementation(() => []);
  });

  afterEach(() => {
    useDispatchMock.mockClear();
    useSelectorMock.mockClear();
    cleanup();
  });

  const mockStore = {
    main: {
      allowIdentifyElements: true,
    },
    pageObject: {
      currentPageObject: "1111",
    },
    locators: {
      status: identificationStatus.success,
      ids: [],
      entities: {},
    }
  };

  test("renders empty page object", () => {
    const { container } = render(<PageObjList />);
    expect(container).toMatchSnapshot();
  });
});
