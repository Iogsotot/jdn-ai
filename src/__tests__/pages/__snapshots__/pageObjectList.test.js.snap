// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pageObjectList component renders empty page object 1`] = `
<div>
  <div
    class="jdn__locatorsList"
  >
    <div
      class="jdn__locatorsList-header"
    >
      <span
        class="jdn__locatorsList-header-title"
      >
        Page Objects
      </span>
      <div
        class="jdn__locatorsList-header-buttons"
      >
        <button
          class="ant-btn ant-btn-default jdn__buttons"
          type="button"
        >
          <span
            class="anticon"
            role="img"
          />
          <span>
            New page object
          </span>
        </button>
        <button
          class="ant-btn ant-btn-default ant-btn-dangerous"
          data-testid="remove-button"
          type="button"
        >
          <span
            class="anticon"
            fill="#D82C15"
            role="img"
          />
        </button>
        <button
          class="ant-btn ant-btn-primary"
          type="button"
        >
          <span
            class="anticon"
            fill="#c15f0f"
            role="img"
          />
        </button>
      </div>
    </div>
    <div
      class="jdn__locatorsList-content jdn__pageObj-content"
    >
      <div
        class="ant-collapse ant-collapse-icon-position-left"
      >
        <div
          class="ant-collapse-item ant-collapse-item-active"
        >
          <div
            aria-expanded="true"
            class="ant-collapse-header"
            role="button"
            tabindex="0"
          >
            <div>
              <span
                class="anticon ant-collapse-arrow"
                fill="#808080"
                role="img"
              />
            </div>
            <span
              class="anticon jdn__locatorsList-status"
              role="img"
            />
            pageObject1
            <div
              class="ant-collapse-extra"
            >
              <a
                data-testid="dropdown-button"
              >
                <span
                  class="anticon ant-dropdown-trigger"
                  role="img"
                  tabindex="-1"
                />
              </a>
            </div>
          </div>
          <div
            class="ant-collapse-content ant-collapse-content-active"
          >
            <div
              class="ant-collapse-content-box"
            >
              <div
                class="jdn__pageObject__settings"
              >
                <div
                  class="jdn__pageObject__settings-url"
                />
                <div
                  class="jdn__generationButtons"
                >
                  <div
                    class="ant-space ant-space-vertical"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-bottom: 16px;"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <label
                            class=""
                            for="library"
                          >
                            Library:
                          </label>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <div
                            class="ant-select jdn__select ant-select-single ant-select-show-arrow"
                          >
                            <div
                              class="ant-select-selector"
                            >
                              <span
                                class="ant-select-selection-search"
                              >
                                <input
                                  aria-activedescendant="library_list_0"
                                  aria-autocomplete="list"
                                  aria-controls="library_list"
                                  aria-haspopup="listbox"
                                  aria-owns="library_list"
                                  autocomplete="off"
                                  class="ant-select-selection-search-input"
                                  id="library"
                                  readonly=""
                                  role="combobox"
                                  style="opacity: 0;"
                                  type="search"
                                  unselectable="on"
                                  value=""
                                />
                              </span>
                              <span
                                class="ant-select-selection-item"
                                title="Material UI"
                              >
                                Material UI
                              </span>
                            </div>
                            <span
                              aria-hidden="true"
                              class="ant-select-arrow"
                              style="user-select: none;"
                              unselectable="on"
                            >
                              <span
                                aria-label="down"
                                class="anticon anticon-down ant-select-suffix"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-primary jdn__buttons"
                            type="button"
                          >
                            <span
                              aria-label="search"
                              class="anticon anticon-search"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="search"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                                />
                              </svg>
                            </span>
                            <span>
                              Generate
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
